src/v/resource_mgmt/available_memory.h

```diff
@@ -0,0 +1,143 @@
+     * Available memory is free memory, plus all the memory that can be in
+     * princple reclaimed from subsystems which have registered a reclaim hook.
+     *
+     * Effectively, this is get_reclaimable() +
+     * ss::memory::stats().free_memory()
+     *
+     * @return size_t the estimated amount of available memory in the system
+     */
+    size_t get_available() const;
+
+    /**
+     * This method calculates the current reclaimable by polling all the
+     * registered reclaimers for how much reclaimable memory they hold, and
+     * returning the sum of all of these amounts.
+     *
+     * The returned value is the assumed amount of memory that could be released
+     * back to the allocator if all reclaimers did an exhaustive reclaim.
+     */
+    size_t get_reclaimable() const;
+
+    /**
+     * @brief Register associated metric(s) for available memory.
+     *
+     * This call registers an "available_memory" metric in the default
+     * metric groups, which reports the get_available() call when it is
+     * polled. Without this call, the class still works fine, exposing
+     * its API to callers, but no metrics are created.
+     */
+    void register_metrics();
+
+    /**
+     * @brief Register an available memory reporter.
+     *
+     * A reporter is a component that can reclaim memory, and so needs to
+     * report the amount of reclaimable memory it holds.
+     *
+     * This method returns a deregister_holder, which when destroyed
+     * deregisters this reporter. Typically, this object will be stored as
+     * a memory of the object to which the reporter function points so that
+     * the function does not access a dangling reference after it is destroyed.
+     *
+     * @param name name of the reporter
+     * @param reporter_fn a function which returns the amount of reclaimable
+     * memory held at any moment in time by this reporter
+     * @return deregister_holder an object that deregisters the reporter
+     */
+    template<typename F>
+    deregister_holder
+    register_reporter(const ss::sstring& name, F reporter_fn) {
+        return inner_register(name, std::move(reporter_fn));
+    }
+
+    /**
+     * @brief Get a reference to the shard-global available_memory instance.
+     *
+     * @return The global available_memory object for this shard.
+     */
+    static available_memory& local() { return _local_instance; }
+
+private:
+    using handle_type = deregister_holder::handle_type;
+
+    // type erased function to report available memory for a
+    // registered reclaimer
+    using afn = std::function<size_t()>;
+
+    struct reporter {
+        ss::sstring name;
+        afn avail_fn;
+        handle_type handle;
+    };
+
+private:
+    friend deregister_holder;
+
+    deregister_holder inner_register(const ss::sstring& name, afn&& fn);
+
+    void deregister(handle_type handle);
+
+    static thread_local available_memory _local_instance;
+
+    std::vector<reporter> _reporters;
```

Comment:
nit: it may be easier to deal with registering/unregistering the handlers when using `boost::intrusive_list` this way a registration holder (having an intrusive list hook) would automatically unlink itself from the list when destroyed.

Reply:
@mmaslankaprv thank you very much for the suggestion: I have done this in the recent upload and it simplifies thing a lot. 

Now there is no separate `deregister_holder` class, but just a `unique_ptr<reporter>` passed back and `reporter` is auto-unlink so it will remove itself.